core/src/main/java/hudson/Util.java

```diff
@@ -507,39 +511,23 @@ public static boolean isSymlink(@Nonnull File file) throws IOException {
          *  both as symlinks, let's use that function and always call it before calling down to the
          *  NIO2 API.
          *
+         *  JENKINS-39179, JENKINS-36088.
+         *  JNA can cause deadlocks, so we avoid it. WindowsFileAttributes.isOther() includes devices,
+         *  but reading the attributes of a device with NIO fails or returns false for isOther(). Named
+         *  pipes are not devices, and return false for isOther().
          */
-        if (Functions.isWindows()) {
-            try {
-                return Kernel32Utils.isJunctionOrSymlink(file);
-            } catch (UnsupportedOperationException | LinkageError e) {
-                // fall through
-            }
-        }
-        Boolean r = isSymlinkJava7(file);
-        if (r != null) {
-            return r;
-        }
-        String name = file.getName();
-        if (name.equals(".") || name.equals(".."))
-            return false;
-
-        File fileInCanonicalParent;
-        File parentDir = file.getParentFile();
-        if ( parentDir == null ) {
-            fileInCanonicalParent = file;
-        } else {
-            fileInCanonicalParent = new File( parentDir.getCanonicalPath(), name );
-        }
-        return !fileInCanonicalParent.getCanonicalFile().equals( fileInCanonicalParent.getAbsoluteFile() );
-    }
-
-    @SuppressFBWarnings("NP_BOOLEAN_RETURN_NULL")
-    private static Boolean isSymlinkJava7(@Nonnull File file) throws IOException {
         try {
             Path path = file.toPath();
-            return Files.isSymbolicLink(path);
-        } catch (Exception x) {
-            throw (IOException) new IOException(x.toString()).initCause(x);
+            BasicFileAttributes attrs = Files.readAttributes(path, BasicFileAttributes.class, LinkOption.NOFOLLOW_LINKS);
+            if (attrs.isSymbolicLink()) {
+                return true;
+            } else if (attrs instanceof DosFileAttributes) {
+                return attrs.isOther();
+            } else {
+                return false;
+            }
+        } catch (NoSuchFileException e) {
+            return false;
```

Comment:
Files.isSymbolicLink() swallows any `IOException` thrown by Files.readAttributes(). `NoSuchFileException` is thrown regularly by the test suite. Should I catch `AccessDeniedException` here as well, or `IOException` to keep the same behavior as before?

We could also fall back to the `getCanonicalFile` tests if an exception is thrown here.

Reply:
The behavior as implemented seems appropriate. If called on a nonexistent file, we return false; if anything else goes wrong, we throw an exception. Right?

Reply:
Yes it seems reasonable to me, I was just surprised that `Files.isSymbolicLink` swallows everything and wasn't sure if there was a good reason to do the same here.

Reply:
Not sure what the motivation was. Swallowing all exceptions and collapsing to a boolean return value seems like a poor design to me. When there is a particular exception you foresee being thrown for a particular reason, then it is fine to handle it accordingly (as you do here).